File: output/qwen3/c/Struct/deposit.c
#include "stdlib.h"

/*@ predicate account(struct account *p; int balance) =
    p != 0 &*&
    malloc_block_account(p, sizeof(struct account)) &*&
    p->balance |-> balance;
@*/

struct account
{
    int balance;
};

struct account *create_account()

{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    //@ assume myAccount != 0;
    //@ leak malloc_block_account(myAccount, sizeof(struct account));
    myAccount->balance = 0;
    //@ close account(myAccount, 0);
    return myAccount;
}

void account_deposit(struct account *myAccount, int amount)

{
    //@ open account(myAccount, ?b);
    myAccount->balance += amount;
    //@ close account(myAccount, b + amount);
}

int account_get_balance(struct account *myAccount)

{
    //@ open account(myAccount, ?b);
    int result = myAccount->balance;
    //@ close account(myAccount, b);
    return result;
}

void account_set_balance(struct account *myAccount, int newBalance)

{
    //@ open account(myAccount, ?_);
    myAccount->balance = newBalance;
    //@ close account(myAccount, newBalance);
}

void account_dispose(struct account *myAccount)

{
    //@ open account(myAccount, ?_);
    free(myAccount);
}

int main()

{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output_at_5/gpt52/sample_2/rust/rust/Struct/deposit.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Account {
    balance: i32,
}

/*@

pred account(p: *mut Account; bal: i32) =
    malloc_block_Account(p) &*&
    (*p).balance |-> bal;

@*/

impl Account {

    unsafe fn create() -> *mut Account
    //@ req true;
    //@ ens account(result, 0);
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        //@ close account(my_account, _);
        (*my_account).balance = 0;
        //@ open account(my_account, _);
        //@ close account(my_account, 0);
        return my_account;
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32
    //@ req account(my_account, ?bal);
    //@ ens account(my_account, bal) &*& result == bal;
    {
        //@ open account(my_account, bal);
        let res = (*my_account).balance;
        //@ close account(my_account, bal);
        return res;
    }

    unsafe fn set_balance(my_account: *mut Account, new_balance: i32)
    //@ req account(my_account, _);
    //@ ens account(my_account, new_balance);
    {
        //@ open account(my_account, _);
        (*my_account).balance = new_balance;
        //@ close account(my_account, new_balance);
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32)
    //@ req account(my_account, ?bal);
    //@ ens account(my_account, bal + amount);
    {
        //@ open account(my_account, bal);
        (*my_account).balance += amount;
        //@ close account(my_account, bal + amount);
    }

    unsafe fn dispose(my_account: *mut Account)
    //@ req account(my_account, _);
    //@ ens true;
    {
        //@ open account(my_account, _);
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }

}

fn main()
    //@ req true;
    //@ ens true;
{
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        //@ assert b == 15;
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}